### PR TITLE
feat(vrl): Add `decode_mime_q` function

### LINKED ISSUE #14325: feat(vrl): Add Q-encoding decoding function
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

To be able to decode q encoded strings as defined in https://datatracker.ietf.org/doc/html/rfc2047 section 2.

### Attempted Solutions

It's possible to implement this decoder in lua but it's suboptimal from performance side and is also bug prone. 

An example lua code:
```lua
function hdr_decode(s)
  local i, j = s:lower():find("=?utf-8?q?", 1, true);
  if i then
      local k, l = s:find("?=", j, true);
      local s_ = s
          :sub(j+1, k-1)
          :gsub("_", " ")
          :gsub("=([a-fA-F0-9][a-fA-F0-9])", function(c) return string.char(tonumber(c, 16)) end);
      return hdr_decode(s:sub(1, i-1) .. s_ .. s:sub(l+1))
  end

  local i, j = s:lower():find("?q?", 1, true);
  if i then
      local s_ = s
          :sub(1+j)
          :gsub("_", " ")
          :gsub("=([a-fA-F0-9][a-fA-F0-9])", function(c) return string.char(tonumber(c, 16)) end);
      return hdr_decode(s_)
  end

  i, j = s:lower():find("=?utf-8?b?", 1, true);
  if i then
      local k, l = s:find("?=", j, true);
      local s_ = s:sub(j+1, k-1):gsub("[%w%+/][%w%+/][%w%+/=][%w%+/=]",
          function(w)
              local digits = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
              local a = digits:find(w:sub(1, 1), 1, true);
              local b = digits:find(w:sub(2, 2), 1, true);
              local c = digits:find(w:sub(3, 3), 1, true);
              local d = digits:find(w:sub(4, 4), 1, true);
              return string.char(
                  (a-1)*4 + math.floor((b-1)/16),
                  (b-1)%16*16 + math.floor(((c or 1)-1)/4),
                  ((c or 1)-1)%4*64 + ((d or 1)-1)
              ):sub(1, d and 3 or c and 2 or 1);
          end
      );
      return hdr_decode(s:sub(1, i-1) .. s_ .. s:sub(l+1));
  end

  return s;
end
```

### Proposal

Add new `decode_q` function to VRL implementing rfc2047 specification. The function would accept q encoded string and output decoded utf8 strings and fail if input string isn't q-encoded. A slight relaxation of specification of making charset optional and defaulting to utf-8 in that case would be useful.

Useful libraries:
* https://crates.io/crates/quoted_printable
* https://crates.io/crates/encoding_rs

### References

_No response_

### Version

vector 0.25.0

### PR DESCRIPTION
**Pull request by ktff**
Closes  #14325

cc. @jszwedko, @sim0nx

### CODE REVIEW

File: lib/vrl/stdlib/src/decode_q.rs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        let charset = self.charset.unwrap_or("utf-8");
+        let charset = Charset::for_label_no_replacement(charset.as_bytes())
+            .ok_or_else(|| format!("Unable to decode {:?} value", charset))?;
+        let (cow, _) = charset.decode_without_bom_handling(&decoded);
+        Ok(cow.into_owned())
+    }
+}
+
+#[cfg(test)]
+mod test {
```
Comment (neuronull) [inline lib/vrl/stdlib/src/decode_q.rs:191]: Thanks for adding tests!
Would it be possible to add an error path testing to cover some of the errors possible from calling `decode_q()`  / `decode_word()` ?

**Review by neuronull — COMMENTED** [review]
Thanks for the contribution!

This looks alright to me, but I'll see if we can get someone with more VRL experience to chime in.

**Comment by fuchsnj** [comment]
The VRL related code looks fine. I'm not familiar with "q encoding", but I do wonder if there is a better name we could give the function. `encode_q` sounds rather generic.

**Comment by neuronull** [comment]
> The VRL related code looks fine. I'm not familiar with "q encoding", but I do wonder if there is a better name we could give the function. `encode_q` sounds rather generic.

While I don't disagree, I'm not sure I could think of a better name :thinking: 

I found this: https://www.rfc-editor.org/rfc/rfc2047#section-4.2

**Comment by JeanMertz** [comment]
> While I don't disagree, I'm not sure I could think of a better name 🤔

Ruby has them under the [`Mail` type](https://www.rubydoc.info/github/mikel/mail/Mail%2FEncodings%2Eq_value_encode) and Perl has them under the [`Mime` type](https://perldoc.perl.org/5.8.1/Encode::MIME::Header).

I think it makes sense to either call this `decode_mime_q` or `decode_mail_q`, I'd probably go for the former.

**Comment by JeanMertz** [comment]
Also, I believe this would be the first decoding function that doesn't have the opposite encoding variant. Could we add that in this PR as well?

See: https://vrl.dev/functions/#codec-functions

**Comment by jszwedko** [comment]
👍 to `decode_mime_q`.

File: lib/vrl/stdlib/src/decode_q.rs (hunk 1 above)
Comment (ktff) [reply lib/vrl/stdlib/src/decode_q.rs:191]: Added, and also fixed a case of no encoded-word.

**Comment by ktff** [comment]
Renamed to `decode_mime_q`.

**Comment by ktff** [comment]
@JeanMertz 
> Also, I believe this would be the first decoding function that doesn't have a matching encoding variant. Could we add that in this PR as well?

Encoding variant isn't in the scope nor do I think it would be useful to have it. Since this encoding was made to avoid breaking message handling software that existed in 1996, I don't see/think there is a demand for it.

File: lib/vrl/stdlib/src/decode_q.rs (hunk 1 above)
Comment (neuronull) [reply lib/vrl/stdlib/src/decode_q.rs]: Thanks!

File: lib/vrl/stdlib/src/decode_mime_q.rs [hunk 2]
```diff
@@ -37,67 +37,70 @@ impl Function for DecodeQ {
...
 
                 Ok(result)
             },
         ),
-        map(parse_internal_q, |word| word.decode_word()),
+        alt((
+            map_opt(parse_internal_q, |word| {
+                word.decode_word().map(|word| Ok(word)).ok()
+            }),
+            success(Ok(String::new())),
```
Comment (neuronull) [inline lib/vrl/stdlib/src/decode_mime_q.rs:102]: Looks like CI has flagged this as having a redundant closure. `make check-clippy` should show you it locally.

**Review by neuronull — COMMENTED** [review]
Thanks for doing the rename and adding the extra test cases!

One clippy issue I flagged.

Otherwise this looks good to me. I'll let @JeanMertz comment on the encode line of discussion.

**Comment by JeanMertz** [comment]
> Encoding variant isn't in the scope

This isn't particularly convincing. My argument is that we should make it in scope, given the facts I laid out above.

> nor do I think it would be useful to have it. Since this encoding was made to avoid breaking message handling software that existed in 1996, I don't see/think there is a demand for it.

This _is_ a more convincing argument. VRL is still meant to allow people to remap their data, which_might_ include remapping a `q` encoded blob and then passing it on as-is (e.g. re-encode it again).

Having said that, I do agree that use-case might be less relevant, so I'm also fine with merging this after we create a tracking issue to gauge interest in a potential `encode_mime_q` function in the future 👍

**Comment by ktff** [comment]
@JeanMertz I've created #14992 for tracking interest in `encode_mime_q`.

**Review by JeanMertz — APPROVED** [review]

File: lib/vrl/stdlib/src/decode_mime_q.rs [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+impl<'a> EncodedWord<'a> {
+    fn decode_word(&self) -> Result<String> {
+        // Modified version from https://github.com/staktrace/mailparse/blob/a83d961fe53fd6504d75ee951a0e91dfea03c830/src/header.rs#L39
+
+        // Decode
+        let decoded = match self.encoding {
+            "B" | "b" => BASE64_MIME
+                .decode(self.input.as_bytes())
+                .map_err(|_| "Unable to decode base64 value")?,
```
Comment (JeanMertz) [inline lib/vrl/stdlib/src/decode_mime_q.rs:165]: Does this mean we support both "b" and "q" decoding in this function?

Comment (JeanMertz) [reply lib/vrl/stdlib/src/decode_mime_q.rs:165]: Ah, I see the Cue docs call this out:

> Replaces q-encoded or base64-encoded

It's a bit weird, but it's a weird/obscure encoding anyway, so this seems fine to me 👍

Comment (ktff) [reply lib/vrl/stdlib/src/decode_mime_q.rs:165]: Yep it's weird. Even that base64 encoding is special version for MIME.

**Comment by ktff** [comment]
Test `Confirm AWS credentials are minimally valid` keeps failing with error:
> Error: Credentials could not be loaded, please check your action inputs: Could not load credentials from any providers

Not sure why.

**Comment by spencergilbert** [comment]
> Test `Confirm AWS credentials are minimally valid` keeps failing with error:
> 
> > Error: Credentials could not be loaded, please check your action inputs: Could not load credentials from any providers
> 
> Not sure why.

👍 expected for the time being, not a blocker!

**Review by neuronull — APPROVED** [review]

**Comment by neuronull** [comment]
Thanks @ktff  !!

**Pull request merged**